lms/services/oauth_http.py

```diff
@@ -66,6 +99,42 @@ def access_token_request(self, token_url, redirect_uri, auth, authorization_code
             },
         )
 
+    def refresh_token_request(self, token_url, redirect_uri, auth):
+        """
+        Make a refresh token request and save the new token in the DB.
+
+        Send an OAuth 2.0 "refresh token request"
+        (https://datatracker.ietf.org/doc/html/rfc6749#section-6) to get a new
+        access token for the current user and save it to the DB.
+
+        :raise OAuth2TokenError: if we don't have a refresh token for the user
+        :raise HTTPError: if the HTTP request fails
+        :raise ValidationError: if the server's access token response is invalid
+        """
+        refresh_token = self._oauth2_token_service.get().refresh_token
+
+        try:
+            return self._token_request(
+                token_url=token_url,
+                auth=auth,
+                data={
+                    "redirect_uri": redirect_uri,
+                    "grant_type": "refresh_token",
+                    "refresh_token": refresh_token,
+                },
+            )
+        except HTTPError as err:
+            try:
+                error_dict = _OAuthAccessTokenErrorResponseSchema(err.response).parse()
+            except ValidationError:
+                pass
+            else:
+                if error_dict["error"] == "invalid_grant":
+                    # Looks like our refresh token has expired or been revoked.
+                    raise OAuth2TokenError() from err
```

Comment:
If we _can_ parse the error response body and identify it as an `"invalid_grant"` error then do so because we want to raise `OAuth2TokenError` so that the frontend shows the user the **Authorize** dialog rather than an error dialog. If not just discard the `ValidationError` and re-raise the original `HTTPError` which will show a **Something went wrong** dialog to the user.